plugin/kubernetes: error NXDOMAIN for TXT lookups

LAMRobinson opened the pull request:
### 1. Why is this pull request needed and what does it do?
All non "dns-version" TXT queries should return a NXDOMAIN rather than NOERROR

### 2. Which issues (if any) are related?
#5736

### 3. Which documentation changes (if any) need to be made?

### 4. Does this introduce a backward incompatible change or deprecation?

robinsla-deshaw committed:
plugin/kubernetes: error NXDOMAIN for TXT lookups

On all non "dns-version" lookups return a valid error status

Signed-off-by: Laurence Robinson <laurence_robinson@live.co.uk>

chrisohaver commented on plugin/kubernetes/kubernetes.go:
```diff
@@ -102,10 +102,10 @@ func (k *Kubernetes) Services(ctx context.Context, state request.Request, exact
 		segs := dns.SplitDomainName(t)
 		if len(segs) != 1 {
-			return nil, nil
+			return nil, errInvalidRequest
```
The answer should be the same as in the next `if`.
This and the following `if` can be combined into one.

LAMRobinson replied:
I ended up restructuring this a bit, rather than checking for a negative (make sure this isn't a query for "dns-version"), we now check if it is a query for "dns-version" and serve that if so, otherwise return NXDOMAIN

chrisohaver commented on plugin/kubernetes/kubernetes.go:
```diff
@@ -102,10 +102,10 @@ func (k *Kubernetes) Services(ctx context.Context, state request.Request, exact
 		}
 		if segs[0] != "dns-version" {
-			return nil, nil
+			return nil, errNoItems
```
Strictly speaking, in the case that the TXT query is for a name that exists for another record type, the answer should be "NODATA".  This is expected DNS behavior.

So, before returning `errNoItems` (which results in NXDOMAIN), this code would need to check to see if a service with the name exists.  If one does exist, then the error returned should be `nil` (which results in NODATA).

LAMRobinson replied:
Good point - I've got this working:
```
$ dig foo.svc.prod.cluster.local TXT @192.168.240.10 | grep status
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 49658

$ dig kubernetes.default.svc.prod.cluster.local TXT @192.168.240.10 | grep status
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 15694
```

chrisohaver reviewed (changes requested):
Thanks! Comments in line. Please also add test cases.

robinsla-deshaw committed:
add tests

Signed-off-by: Laurence Robinson <laurence_robinson@live.co.uk>

robinsla-deshaw committed:
Adjust TXT record logic to follow RFC

 * If a FQDN exists for another record type return NOERROR
 * If not return NXDOMAIN

Signed-off-by: Laurence Robinson <laurence_robinson@live.co.uk>

chrisohaver reviewed (commented):
Looks good.  Can you sign the commits?

robinsla-deshaw committed:
don't propegate service population errors

all we care about is if services are empty, actual errors due to the
query are irrelevant as we end up at an error state anyway

Signed-off-by: Laurence Robinson <laurence_robinson@live.co.uk>

robinsla-deshaw committed:
fix tabs/spaces

Signed-off-by: Laurence Robinson <laurence_robinson@live.co.uk>

LAMRobinson committed:
Support dns-version.<zone> TXT only

Had a regression where the code allowed dns-version.*.<zone>

Signed-off-by: Laurence Robinson <laurence_robinson@live.co.uk>

chrisohaver commented on plugin/kubernetes/kubernetes.go:
```diff
@@ -110,7 +110,7 @@ func (k *Kubernetes) Services(ctx context.Context, state request.Request, exact
 
 		// Hard code the only valid TXT - "dns-version.<zone>"
```
I think this check should happen before the `Records` check, so the more expensive Records check is only performed for TXT queries that do not match the static dns-version record.

LAMRobinson replied:
good idea, done

chrisohaver reviewed (commented):
Thanks. Please also fix the reviewdog whitespace notifications, imbedded in-line in Files Changed.

LAMRobinson committed:
Fix whitespace / move around logic

Signed-off-by: Laurence Robinson <laurence_robinson@live.co.uk>

LAMRobinson committed:
Fix whitespace..

Signed-off-by: Laurence Robinson <laurence_robinson@live.co.uk>

LAMRobinson committed:
fix test

Signed-off-by: Laurence Robinson <laurence_robinson@live.co.uk>

chrisohaver commented on plugin/kubernetes/handler_test.go:
```diff
@@ -251,7 +251,9 @@ var dnsTestCases = []kubeTestCase{
-		Answer: []dns.RR{},
+		Answer: []dns.RR{
+			test.SOA("cluster.local.	5	IN	SOA	ns.dns.cluster.local. hostmaster.cluster.local. 1499347823 7200 1800 86400 5"),
+        },
```
This doesn't look right.

For a NODATA response, the SOA record should be in the Authority section (`Ns`).

LAMRobinson replied:
yeah I expected as such, but was trying to troubleshoot why a test was failing - ill revert

LAMRobinson committed:
fix test

Signed-off-by: Laurence Robinson <laurence_robinson@live.co.uk>

LAMRobinson committed:
fix bad testcase

Signed-off-by: Laurence Robinson <laurence_robinson@live.co.uk>

LAMRobinson committed:
return SOA on missing TXT test case

Signed-off-by: Laurence Robinson <laurence_robinson@live.co.uk>

chrisohaver reviewed (changes requested):
Looks good.  Please fix line 253 per the reviewdog comment in handler_test.go, and it should be good to merge.

LAMRobinson committed:
clean up whitespace

Signed-off-by: Laurence Robinson <laurence_robinson@live.co.uk>
